### PR TITLE
Extend player profile API to support skin changes

### PR DESCRIPTION
**Pull request by NickAcPT**
Added code that refreshes the player's skin by sending packets with a special order, telling the client to respawn the player and re-apply the game profile

### CODE REVIEW

File: Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+-        }
++        refreshPlayer();
+     }
+     public PlayerProfile getPlayerProfile() {
+         return new CraftPlayerProfile(this).clone();
+     }
++
++    private void refreshPlayer() {
++        //Calculate the correct dimension
++        int dimension = (int) ((WorldServer)getHandle().getWorld()).getDimension();
```
Comment (kashike) [inline Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch:98]: Where do these magic numbers come from?

File: Spigot-API-Patches/0126-Extend-player-profile-API-to-support-skin-changes.patch [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+diff --git a/src/main/java/com/destroystokyo/paper/profile/PlayerProfile.java b/src/main/java/com/destroystokyo/paper/profile/PlayerProfile.java
+index e060c38a..98eade06 100644
+--- a/src/main/java/com/destroystokyo/paper/profile/PlayerProfile.java
++++ b/src/main/java/com/destroystokyo/paper/profile/PlayerProfile.java
+@@ -132,6 +132,14 @@ public interface PlayerProfile {
+      * @return If the profile is now complete (has UUID and Name) (if you get rate limited, this operation may fail)
+      */
+     boolean complete(boolean textures);
++    /**
```
Comment (aikar) [inline Spigot-API-Patches/0126-Extend-player-profile-API-to-support-skin-changes.patch:16]: Based on what the impl is below, i'm not a fan of adding this. Servers should have their desired online mode set correctly using the Paper Bungee Online mode setting.

Comment (aikar) [reply Spigot-API-Patches/0126-Extend-player-profile-API-to-support-skin-changes.patch:16]: Plus the javadocs are a copy and don't represent what its for.

File: Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+     public boolean complete(boolean textures) {
++        return complete(textures, false);
++    }
++
++    public boolean complete(boolean textures, boolean force) {
+         MinecraftServer server = MinecraftServer.getServer();
+ 
+         boolean isOnlineMode = server.getOnlineMode() || (SpigotConfig.bungee && PaperConfig.bungeeOnlineMode);
+         boolean isCompleteFromCache = this.completeFromCache(true);
+-        if (isOnlineMode && (!isCompleteFromCache || textures && !hasTextures())) {
```
Comment (aikar) [inline Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch:24]: See API review, this is unneeded. isOnlineMode is true if the server is on offline mode but has the paper online bungee mode turned on.

File: Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch (hunk 1 above)
Comment (aikar) [reply Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch:98]: This should just be 

```java
entityplayer.playerConnection.sendPacket(new PacketPlayOutRespawn(entityplayer.dimension, entityplayer.world.getDifficulty(), entityplayer.world.getWorldData().getType(), entityplayer.playerInteractManager.getGameMode()));
```

As done in PlayerList.a(EntityPlayer, int);

Why was all this magical mapping done? These values seem to hold no meaning.

File: Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
++        if (dimension == 1 || dimension == 14)
++            dimension = -1;
++        if (dimension == 16 || dimension == 12)
++            dimension = 1;
++
++        //Respawn packet
++        Packet respawn = new PacketPlayOutRespawn(dimension, ((WorldServer)getHandle().getWorld()).worldData.getDifficulty(), ((WorldServer)getHandle().getWorld()).worldData.getType(), EnumGamemode.getById(getHandle().playerInteractManager.getGameMode().getId()));
++
++        Location l = getLocation();
++
```
Comment (aikar) [inline Spigot-Server-Patches/0336-Extend-player-profile-API-to-support-skin-changes.patch:110]: All of this should be unnecessary. We retrack the entity already to other players, which picks up the changes. The scope of this PR is trying to solve the player seeing their own change.

Why would the player need help seeing new equipment? It's just the profile we are trying here.

Shouldn't the respawn packet be the main need.

**Pull request closed without merging**